### PR TITLE
feat(explore): Move chart header to top of the page

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
This PR implements part of chart header redesign (final design available here: https://github.com/apache/superset/discussions/19099)

List of changes:
- move chart header to top of the page
- change looks and behaviour of editable title - input grows when user types text, no wrapping, when title overflows it gets truncated and tooltip with full name appears on hover
- when title is truncated, clicking (enabling editing mode) moves the cursor to the end of the title
- Applied additional paddings to Dataset label and column search on the left so that everything is aligned
- Deleting the whole title is allowed - we don't revert to the last title when user clears the input. Instead, a placeholder is shown
- When new chart is created, the default title is empty instead of "- untitled", which means that we display the placeholder
- Editing the name of existing chart gets recorded in "Altered" table
- if user doesn't have edit permissions, instead of an input element display span element that looks and behaves the same except for editing capabilities

Next steps:
- move timer, rows retrieved and cached labels to chart container (implemented in https://github.com/apache/superset/pull/19458)
- move save button to the right of chart header 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

https://user-images.githubusercontent.com/15073128/161726241-ffdf4019-4904-4ef0-8efd-b0fa749fdd36.mov

### TESTING INSTRUCTIONS
1. Open a chart that you have permissions to edit
2. Click the title, verify that cursor appears
3. When there's no text a placeholder should appear
4. When you start typing, the input's width should adjust to its content
5. When input is wider than the header container, it should move to the position of cursor if in edit mode, or be truncated if not in edit mode
6. If not editing and title is truncated, display tooltip with full title on hover 

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API
CC @kasiazjc

### CODE REVIEW

File: superset-frontend/src/explore/components/ExploreChartHeader/ChartEditableTitle/index.tsx [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+          <span
+            className="chart-title"
+            aria-label={t('Chart title')}
+            ref={contentRef}
+          >
+            {currentTitle}
+          </span>
+        )}
+      </Tooltip>
+      <span ref={sizerRef} className="input-sizer" aria-hidden tabIndex={-1} />
```
Comment (kgabryje) [inline superset-frontend/src/explore/components/ExploreChartHeader/ChartEditableTitle/index.tsx:196]: That span is invisible, we need it to measure input width in order to scale it's width dynamically

File: superset-frontend/src/explore/components/ExploreChartHeader/index.jsx [hunk 2]
```diff
@@ -235,8 +238,18 @@ export class ExploreChartHeader extends React.PureComponent {
...
     );
     return (
-      <StyledHeader id="slice-header" className="panel-title-large">
+      <StyledHeader id="slice-header">
         <div className="title-panel">
+          <ChartEditableTitle
+            title={this.getSliceName()}
+            canEdit={
+              !slice ||
+              this.props.can_overwrite ||
```
Comment (villebro) [inline superset-frontend/src/explore/components/ExploreChartHeader/index.jsx:247]: nit: while we're using the destructured `slice`, could we also destructure this like the others on line 224: `const { ..., can_overwrite: canOverwrite } = this.props`. Also maybe `this.props.actions.updateChartTitle`

**Review by villebro — APPROVED** [review]
Approving with a minor nit (not important): Very clean and I like this much better than the current UX. A few thoughts that came to mind when testing: it's not obvious to the user that the title has been edited. I believe some time ago it used to save it when pressing return? I'm not saying that was very intuitive, either, but since we have the pill that says when the form data has changed, maybe we should also record the changed title in it?

image

But let's leave that for a follow up, as it's not really directly related to this PR.

**Comment by kgabryje** [comment]
> I believe some time ago it used to save it when pressing return? 

It still does 🙂 I like the idea of including the title in altered table! I'll add that to this PR

File: superset-frontend/src/explore/components/ExploreChartHeader/index.jsx (hunk 2 above)
Comment (kgabryje) [reply superset-frontend/src/explore/components/ExploreChartHeader/index.jsx]: Done!

**Pull request merged**